### PR TITLE
Function editor

### PR DESCRIPTION
**Pull request by NeilFraser**

### CODE REVIEW

File: static/code/editor.js [hunk 1]
```diff
@@ -771,7 +771,181 @@ Code.valueEditor.focus = function(userAction) {
...
+  var source = this.editor_.getValue();
+  var verb = '@delete_prop verb';
+  var dobj = '@delete_prop dobj';
+  var prep = '@delete_prop prep';
+  var iobj = '@delete_prop iobj';
+  if (this.isVerbElement_.checked) {
+    verb = '@set_prop verb = ' + JSON.stringify(this.verbElement_.value);
+    dobj = '@set_prop dobj = ' + JSON.stringify(this.dobjElement_.value);
+    prep = '@set_prop prep = ' + JSON.stringify(this.prepElement_.value);
+    iobj = '@set_prop iobj = ' + JSON.stringify(this.iobjElement_.value);
```
Comment (cpcallen) [inline static/code/editor.js]: Error handling for `JSON.stringify`?

File: demo/core_34_$.www.code.js [hunk 2]
```diff
@@ -230,9 +238,22 @@ $.www.code.editor.www = function(request, response) {
       parts.push(lastPart);
       var selector = $.utils.selector.partsToSelector(parts);
       $.utils.selector.setSelector(value, selector);
+      // Assemble any meta-data for the editor.
+      var meta = '';
+      if (typeof value === 'function') {
+        meta += '// @copy_properties true\n';
+        meta += '// ' + (value.verb ? '@set_prop verb = ' +
+            JSON.stringify(value.verb) : '@delete_prop verb') + '\n';
```
[the remark this answers is not in the document]
Comment (NeilFraser) [reply demo/core_34_$.www.code.js]: Fixed.

File: demo/core_34_$.www.code.js [hunk 3]
```diff
@@ -243,3 +264,26 @@ $.www.code.editor.www = function(request, response) {
...
+  }
+  var metaLines = m[0].split('\n');
+  for (var i = 0; i < metaLines.length; i++) {
+    var meta = metaLines[i];
+    if (meta.match(/^\s*\/\/\s*@copy_properties\s*true\s*$/)) {
+      // @copy_properties true
+        $.utils.transplantProperties(oldValue, newValue);
+    } else if ((m = meta.match(/^\s*\/\/\s*@delete_prop\s+(\S+)\s*$/))) {
+      // @delete_prop dobj
+      delete newValue[m[1]];
```
[the remark this answers is not in the document]
Comment (NeilFraser) [reply demo/core_34_$.www.code.js]: Fixed.

File: demo/core_34_$.www.code.js [hunk 4]
```diff
@@ -243,3 +264,26 @@ $.www.code.editor.www = function(request, response) {
...
+    } else if ((m = meta.match(/^\s*\/\/\s*@set_prop\s+(\S+)\s*=(.+)$/))) {
+      // @set_prop dobj = "this"
+      newValue[m[1]] = JSON.parse(m[2]);
```
[the remark this answers is not in the document]
Comment (NeilFraser) [reply demo/core_34_$.www.code.js]: Fixed.

File: static/code/editor.js (hunk 1 above)
Comment (NeilFraser) [reply static/code/editor.js]: No, this is an HTML input, the value must be a string (or if you put a gun to my head I might be able to figure out how to inject an undefined or similar).  But it definitely can't be Object.create(null), or a recursive data structure.

File: static/code/editor.js [hunk 5]
```diff
@@ -771,7 +771,181 @@ Code.valueEditor.focus = function(userAction) {
...
+Code.functionEditor = new Code.GenericEditor('Function');
+
+/**
+ * Code Mirror editor.  Does not exist until tab is selected.
+ * @type {Object}
+ * @private
+ */
+Code.functionEditor.editor_ = null;
+
+Code.functionEditor.updateDisabled = function() {
```
[the remark this answers is not in the document]
Comment (NeilFraser) [reply static/code/editor.js]: More to the point, this function is called exclusively by createDom, so it should follow.  Fixed.
Also added missing JSDoc.

File: static/code/editor.js (hunk 1 above)
Comment (cpcallen) [reply static/code/editor.js]: Fair point!

**Pull request merged**